### ISSUE TITLE
Basic benchmarks, MacOS

### ISSUE THREAD

**Issue by jondegenhardt** [jondegenhardt/dcat-perf#1]
### Benchmark tests

Four different `dcat` test types were measured, all byLine based. See the `dcat` code for the precise definitions of each test. The GNU `time` function was used to measure results. Four different data files were used representing a spectrum of line widths.

Each test was run five times and the median value recorded. Times in seconds.

A synopsis of the test types:

| Test   |  Input   | Output |
| ----   |   -----  |  ------ |
| byLineInRawOut | std.stdio.File.byLine  |  std.stdio.File.write |
| byLineInBufOut  | std.stdio.File.byLine  | tsv_utils.common.utils.BufferedOutputRange |
| bufferedByLineInBufOut | tsv_utils.common.utils.bufferedByLine | tsv_utils.common.utils.BufferedOutputRange |
| iopipeByLineInBufOut  |  iopipe.byLine | tsv_utils.common.utils.BufferedOutputRange |

### System

MacMini; OS X Mojave, 16 GB RAM, SSD drives.

### Results (LTO/PGO build)

**File:  audio_sai_boxes.txt; 3.4 gb; 18809.4 bytes/line**

| Test  |  elapsed  | user  | system  | memory |
| ----- | --------: | -----: | -------: | -------: |
| byLineInRawOut         |           2.40 |        1.05 |          1.34 |       2800 |
| byLineInBufOut         |           2.60 |        1.24 |          1.36 |       2960 |
| bufferedByLineInBufOut |           1.14 |        0.54 |          0.59 |       2372 |
| iopipeByLineInBufOut   |           1.26 |        0.63 |          0.62 |       2212 |

*Note: Unlike the other files, `audio_sai_boxes.txt` has quite variable length lines, alternating between very short lines (< 20 bytes) and very long lines, typically greater than 20K bytes, up to more than 70K bytes. A file more uniformly 20K bytes might  have different performance characteristics.*

 **File: hepmass_all_train.tsv; 4.8 gb; 739.8 bytes/line**

| Test  |  elapsed  | user  | system  | memory |
| ----- | --------: | -----: | -------: | -------: |
| byLineInRawOut         |           4.40 |        2.52 |          1.87 |       1936 |
| byLineInBufOut         |           3.95 |        2.03 |          1.91 |       1952 |
| bufferedByLineInBufOut |           1.74 |        0.86 |          0.86 |       2084 |
| iopipeByLineInBufOut   |           1.89 |        0.82 |          1.06 |       1972 |

**File: TREE_GRM_ESTN_14mil.tsv; 2.7 gb; 206.0 bytes/line**

| Test  |  elapsed  | user  | system  | memory |
| ----- | --------: | -----: | -------: | -------: |
| byLineInRawOut         |           4.63 |        3.60 |          1.02 |       1936 |
| byLineInBufOut         |           3.13 |        2.07 |          1.05 |       1952 |
| bufferedByLineInBufOut |           1.50 |        1.00 |          0.49 |       2096 |
| iopipeByLineInBufOut   |           1.26 |        0.68 |          0.57 |       1960 |

**File: googlebooks-eng-all-1gram-20120701-s.tsv; 2.2 gb; 21.0 bytes/line**

| Test  |  elapsed  | user  | system  | memory |
| ----- | --------: | -----: | -------: | -------: |
| byLineInRawOut         |          22.49 |       21.63 |          0.83 |       1948 |
| byLineInBufOut         |           9.46 |        8.59 |          0.86 |       1952 |
| bufferedByLineInBufOut |           6.39 |        5.96 |          0.42 |       2096 |
| iopipeByLineInBufOut   |           2.84 |        2.36 |          0.47 |       1972 |

### File details

| Name     |   Size (GB)  |  Bytes  |  Lines    | Bytes/Line |
| -------- | ------: | ------: | ------: | ------: | 
| audio_sai_boxes.txt | 3.4 | 3684160075 | 195868 | 18809.4 |
| hepmass_all_train.tsv | 4.8 | 5178435596 | 7000001 | 739.8 |
| TREE_GRM_ESTN_14mil.tsv | 2.7 | 2883776925 | 14000000 | 206.0 |
| googlebooks-eng-all-1gram-20120701-s.tsv | 2.2 | 2316331839 | 110075424 | 21.0 |

* `audio_sai_boxes.txt` - From UCI Machine Learning Repository, [YouTube Multiview Video Games Data Set](https://archive.ics.uci.edu/ml/datasets/YouTube+Multiview+Video+Games+Dataset). File from the `train` directory in the download.
* `hepmass_all_train.tsv` - The HEPMASS training set from the UCI Machine Learning repository, available [here](https://archive.ics.uci.edu/ml/datasets/HEPMASS). (The `all_train.csv.gz` file. Use `csv2tsv` to convert to TSV.)
* `TREE_GRM_ESTN_14mil.tsv` - From the Forest Inventory and Analysis Database, U.S. Department of Agriculture. The first 14 million lines from the TREE_GRM_ESTN_.csv file, available [here](https://apps.fs.usda.gov/fia/datamart/CSV/datamart_csv.html). (Use `csv2tsv` to convert to TSV.)
* `googlebooks-eng-all-1gram-20120701-s.tsv` - The google one-gram file for the letter 's', available from the [Google Books ngram datasets](http://storage.googleapis.com/books/ngrams/books/datasetsv2.html).

**Comment by dejlek** [comment]
Amazing! Thanks Jon! I hope one day iopipe and std.io get into the Phobos where they rightfully belong (perhaps with some changes in the API)...

**Comment by jondegenhardt** [comment]
@dejlek Agreed, `iopipe` and `std.io` are very promising. Still a few rough spots in the APIs, but this looks very manageable.
